docs/user-defined-functions/index.md

```diff
@@ -0,0 +1,685 @@
+
+```sql
+CREATE TABLE InternationalSalesRecordTable ( 
+    invoice_id STRING, 
+    user_id STRING, 
+    product_id STRING, 
+    quantity STRING, 
+    unit_cost STRING, 
+    `purchase_time` TIMESTAMP(3) METADATA FROM 'timestamp', 
+    WATERMARK FOR purchase_time AS purchase_time - INTERVAL '1' SECOND 
+) WITH ( 
+    'connector' = 'kafka',
+    'topic' = 'flink.international.sales.records', 
+    'properties.bootstrap.servers' = 'my-cluster-kafka-bootstrap.flink.svc:9092', 
+    'properties.group.id' = 'international-sales-record-group', 
+    'value.format' = 'avro-confluent', 
+    'value.avro-confluent.url' = 'http://apicurio-registry-service.flink.svc:8080/apis/ccompat/v6', 
+    'scan.startup.mode' = 'latest-offset'
+);
+```
+
+We can do a simple query to verify that the table was created correctly and that the data is flowing (give it a few seconds to start receiving data):
+
+```sql
+SELECT * FROM InternationalSalesRecordTable;
+```
+
+If that worked, we can now register our UDF as a [temporary catalog function](https://nightlies.apache.org/flink/flink-docs-release-2.0/docs/dev/table/sql/create/#create-function):
+
+```sql
+CREATE TEMPORARY FUNCTION currency_convert
+AS 'com.github.streamshub.CurrencyConverter'
+USING JAR '/opt/currency-converter-1.0-SNAPSHOT.jar';
+```
+
+> Note: Temporary catalog functions [only live as long as the current session](https://nightlies.apache.org/flink/flink-docs-release-2.0/docs/dev/table/functions/overview/#types-of-functions). Provided you have a [Flink catalog](https://nightlies.apache.org/flink/flink-docs-release-2.0/docs/dev/table/catalogs/#catalogs) deployed and configured, you can omit the `TEMPORARY` keyword to create a function that persists across sessions.
+
+> Note: This statement may succeed even if the JAR was not found or has insufficient permissions. You will likely only find this out when you try to use the UDF in a query.
+
+Now, we can use our UDF in a query:
+
+```sql
+SELECT
+    invoice_id,
+    currency_convert(unit_cost) AS iso_unit_cost,
+    unit_cost,
+    quantity
+FROM InternationalSalesRecordTable;
+```
+
+You should start seeing results with both a `unit_cost` field and an `iso_unit_cost` field containing the output of our UDF!
+
+We can also use the UDF in more complex queries e.g. to filter for records with a specific currency and quantity:
+
+```sql
+SELECT 
+    DISTINCT(product_id), 
+    iso_unit_cost, 
+    quantity 
+  FROM (
+      SELECT 
+          invoice_id, 
+          user_id,
+          product_id,
+          CAST(quantity AS INT) AS quantity, 
+          currency_convert(unit_cost) AS iso_unit_cost,
+          unit_cost,
+          purchase_time
+      FROM InternationalSalesRecordTable
+    )
+WHERE 
+    RIGHT(iso_unit_cost, 3) = 'EUR' AND quantity > 1;
+```
+
+> Note: This query might take a while to return results, since there are many currencies used in the data!
```

Comment:
We might need to up the source generators rate as it took 30 or so seconds for this to produce a result.